File: advent_python/day02.py
INPUT = [
    "1-3 a: abcde",
    "1-3 b: cdefg",
    "2-9 c: ccccccccc"
]


def check_passwords(passwords: list) -> int:
    valid_pws = 0
    for pw in passwords:
        pw = pw.strip()
        pw = pw.split(' ')
        pw_str = pw[2]
        min_c = int(pw[0].split('-')[0])
        max_c = int(pw[0].split('-')[1])
        c = str(pw[1][:-1])
        count_c = pw_str.count(c)
        if min_c <= count_c <= max_c:
            valid_pws += 1

    return valid_pws

assert check_passwords(INPUT) == 2


def check_passwords_new(passwords: list) -> int:
    valid_pws = 0
    for pw in passwords:
        pw = pw.strip()
        pw = pw.split(' ')
        pw_str = pw[2]
        pos_1 = int(pw[0].split('-')[0])
        pos_2 = int(pw[0].split('-')[1])
        c = str(pw[1][:-1])
        if ((pw_str[pos_1 - 1] == c) ^ (pw_str[pos_2 - 1] == c)):
            valid_pws += 1

    return valid_pws

assert check_passwords_new(INPUT) == 1

with open('../inputs/day02.txt', 'r') as f:
    data = [l for l in f]
    valid_pws = check_passwords(data)
    print(valid_pws) 
    valid_pws = check_passwords_new(data)
    print(valid_pws)  

File: advent_cpp/day02.cpp
#include <iostream>
#include <fstream>
#include <string>
#include <vector>
#include <algorithm>
#include <cassert>

const std::vector<std::string> INPUT {
    "1-3 a: abcde",
    "1-3 b: cdefg",
    "2-9 c: ccccccccc"
};

std::vector<std::string> read_input(std::string & fileName)
{
    std::vector<std::string> passwords;
    std::ifstream in(fileName.c_str());

    // Check if the object is valid
    if (!in) {
        std::cerr << "Cannot open the file: " << fileName << std::endl;
    }

    std::string str;
    // Read in the file, line by line
    while (std::getline(in, str)) 
    {
        if (str.size() > 0)
            passwords.push_back(str);
    }
    // Close the file
    in.close();

    return passwords;
}

int check_passwords(std::vector<std::string> passwords) 
{
    int valid = 0;
    for (std::string pwd : passwords) 
    {
        // remove newline from the end of the pwd
        pwd.erase(std::remove(pwd.begin(), pwd.end(), '\n'), pwd.end());
        std::size_t p = pwd.find(" ");
        // extract min and max number for character
        std::string min_max = pwd.substr(0, p);
        std::size_t min = std::stoi(min_max.substr(0, min_max.find("-")));
        std::size_t max = std::stoi(min_max.substr((min_max.find("-")+1)));

        // extract the character and the password
        pwd = pwd.substr((p+1));
        char character = pwd.substr(0, pwd.find(":"))[0];
        std::string password = pwd.substr((pwd.find(" ") + 1));
        // count the char occurences
        std::size_t n = std::count(password.begin(), password.end(), character);
        if (n >= min and n <= max) {
            valid++;
        } 

    }
    return valid;
}

int main() 
{   
    std::string fileName {"../inputs/day02.txt"};
    std::vector<std::string> passwords = read_input(fileName);
    std::cout << "Read input successfully." << std::endl;

    assert((check_passwords(INPUT) == 2));
    std::cout << "First test passed!" << std::endl;

    assert((check_passwords(passwords) == 465));
    std::cout << "Second test passed!" << std::endl;
    
    return 0;
}